src/R2N.jl

```diff
@@ -0,0 +1,283 @@
+ψ(s; xₖ) = h(xₖ + s), ‖⋅‖ is a user-defined norm and σₖ > 0 is the regularization parameter.
+The subproblem is solved inexactly by way of a first-order method such as the proximal-gradient
+method or the quadratic regularization method.
+
+### Arguments
+
+* `nlp::AbstractNLPModel`: a smooth optimization problem
+* `h`: a regularizer such as those defined in ProximalOperators
+* `options::ROSolverOptions`: a structure containing algorithmic parameters
+
+The objective, gradient and Hessian of `nlp` will be accessed.
+The Hessian is accessed as an abstract operator and need not be the exact Hessian.
+
+### Keyword arguments
+
+* `x0::AbstractVector`: an initial guess (default: `nlp.meta.x0`)
+* `subsolver_logger::AbstractLogger`: a logger to pass to the subproblem solver (default: the null logger)
+* `subsolver`: the procedure used to compute a step (`PG` or `R2`)
+* `subsolver_options::ROSolverOptions`: default options to pass to the subsolver (default: all defaut options)
+* `selected::AbstractVector{<:Integer}`: (default `1:f.meta.nvar`).
+
+### Return values
+
+* `xk`: the final iterate
+* `Fobj_hist`: an array with the history of values of the smooth objective
+* `Hobj_hist`: an array with the history of values of the nonsmooth objective
+* `Complex_hist`: an array with the history of number of inner iterations.
+"""
+function R2N(
+  f::AbstractNLPModel,
+  h::H,
+  options::ROSolverOptions;
+  x0::AbstractVector = f.meta.x0,
+  subsolver_logger::Logging.AbstractLogger = Logging.NullLogger(),
+  subsolver = R2,
+  subsolver_options = ROSolverOptions(ϵa = options.ϵa),
+  selected::AbstractVector{<:Integer} = 1:(f.meta.nvar),
+) where {H}
+  start_time = time()
+  elapsed_time = 0.0
+  # initialize passed options
+  ϵ = options.ϵa
+  ϵ_subsolver_init = subsolver_options.ϵa
+  ϵ_subsolver = copy(ϵ_subsolver_init)
+  ϵr = options.ϵr
+  Δk = options.Δk
+  verbose = options.verbose
+  maxIter = options.maxIter
+  maxTime = options.maxTime
+  η1 = options.η1
+  η2 = options.η2
+  γ = options.γ
+  α = options.α
+  θ = options.θ
+  β = options.β
+  σmin = options.σmin
+
+  local l_bound, u_bound
+  if has_bounds(f) || subsolver == TRDH
+    l_bound = f.meta.lvar
+    u_bound = f.meta.uvar
+  end
+
+  if verbose == 0
+    ptf = Inf
+  elseif verbose == 1
+    ptf = round(maxIter / 10)
+  elseif verbose == 2
+    ptf = round(maxIter / 100)
+  else
+    ptf = 1
+  end
+
+  # initialize parameters
+  σk = max(1 / options.ν, σmin)
+  xk = copy(x0)
+  hk = h(xk[selected])
+  if hk == Inf
+    verbose > 0 && @info "R2N: finding initial guess where nonsmooth term is finite"
+    prox!(xk, h, x0, one(eltype(x0)))
+    hk = h(xk[selected])
+    hk < Inf || error("prox computation must be erroneous")
+    verbose > 0 && @debug "R2N: found point where h has value" hk
+  end
+  hk == -Inf && error("nonsmooth term is not proper")
+
+  xkn = similar(xk)
+  s = zero(xk)
+  ψ =
+    (has_bounds(f) || subsolver == TRDH) ?
```

Comment:
Do you really want to use TRDH as subsolver? The point of R2N is to be able to compute a prox without a trust region. R2DH would be more appropriate.